### ISSUE TITLE
Elm test should be a first class citizen

### ISSUE THREAD

**Issue by Razzeee** [elm-tooling/elm-language-server#48]

**Comment by andys8** [comment]
Cross link discussion: https://github.com/elm-tooling/elm-language-server/issues/47#issuecomment-498338573

**Comment by Razzeee** [comment]
The real question is how do we achieve this?

**Comment by Razzeee** [comment]
We should probably add a commands for running all tests. At the moment we only run the tests from the file your in on save.

**Comment by andys8** [comment]
We're back to "executing" tests? If so, I think there has to be an opt-out.

It takes 15 seconds for a toy project with fuzz tests. A well tested production project might be even worse.

There is and should be a big difference between making sure, the test is valid elm syntax, and execution the tests. I'm not sure, but it looks like both is done. Is this correct?

Can't really test because of issue file path. See slack :)

**Comment by andys8** [comment]
This is the current result:

```
[Trace - 3:23:19 PM] Received notification 'textDocument/publishDiagnostics'.
Params: {
    "uri": "file:///home/as/dev/repository/vim-emulation/tests/MainTest.elm",
    "diagnostics": [
        {
            "range": {
                "start": {
                    "line": 0,
                    "character": 0
                },
                "end": {
                    "line": 0,
                    "character": 0
                }
            },
            "message": "Equality - Comparison: Expect.equal\n Expected: Insert\n Actual: Normal\n",
            "severity": 1,
            "source": "Elm"
        }
    ]
}
```

There is currently no line number or test name. So it's hard to tell which test failed.

**Comment by Razzeee** [comment]
Those are not provided by elm test unfortunately.

Best would be to add commands to run these. Until the language server spec implement's something test related.

**Comment by andys8** [comment]
It provides the test name as `labels`:

```json
{
  "event": "testCompleted",
  "status": "fail",
  "labels": [
    "MainTest",
    "Main",
    "update",
    "Type a word"
  ],
  "failures": [
    {
      "given": null,
      "message": "Expect.equal",
      "reason": {
        "type": "Equality",
        "data": {
          "expected": "Buffer \"asd1f\"",
          "actual": "Buffer \"asdf\"",
          "comparison": "Expect.equal"
        }
      }
    }
  ],
  "duration": "5"
}
```

But I'm also very much in favor of using commands instead for now.

**Comment by andys8** [comment]
Command would be:

- All: `elm-test` in root project directory
- Specific file: `elm-test tests/MyTest.elm`

**Comment by Razzeee** [comment]
so how would the labels need to be concated to form the name again?

**Comment by andys8** [comment]
It looks like the `labels` are:
* name of the test module
* n `describe` blocks
* name of the test function

**Comment by andys8** [comment]
Intersperse the list with ` > ` should probably look fine.

**Comment by Razzeee** [comment]
Added

**Comment by andys8** [comment]
Do you plan to migrate it to a command instead of on save?

**Comment by Razzeee** [comment]
Probably, but commands are at the moment, what I keep pushing back.

**Comment by andys8** [comment]
Yeah, I try to push back executing tests on save, because I won't be able to upgrade then :D

**Comment by Razzeee** [comment]
Which is confusing to me, as this was already included in the last release

**Comment by andys8** [comment]
I didn't notice because I didn't write new tests in the last weeks ;) At one point I split up a large test file, because it became unusable.

Here is a demo with latest release. It shows, how long it takes to remove/fix a syntax error, because this'll trigger executing the test. The file is already smaller, because I split up the tests, because of lagging at one point.

[![asciicast](https://asciinema.org/a/UylrsxCQpdWvkDNm5IhURpHNL.svg)](https://asciinema.org/a/UylrsxCQpdWvkDNm5IhURpHNL)

**Comment by Razzeee** [comment]
I pushed this WIP branch. https://github.com/elm-tooling/elm-language-server/tree/test-commands

It has two problems:
1. Running all tests -> test runner doesn't say which file it's respond is for https://github.com/rtfeldman/node-test-runner/issues/371
2. Running tests from the current file -> Commands doesn't seem to be able to figure out the current file. Might have overlooked something.

So I would basically rip out test support for the moment.

**Comment by andys8** [comment]
Hey, didn't have a look at the branch, but regarding 1: Is the assumption holding that the first label is always the module name and can therefore deterministicly converted to file name? Is this the case or are there counter examples?
